test/topology_experimental_raft/test_raft_voters.py

```diff
@@ -0,0 +1,90 @@
+#
```

Comment:
The test is fine, but it's hard to judge this single test without more context (the precise requirements, the solution, other tests). I mean that we can merge this test but I don't know how much sense it will make when we merge the rest.

I'll try to explain why I'm a bit worried. I believe the requirements are the following:
- no dc can have at least half of the voters at any point if possible (it's always possible with >= 3 dcs and always impossible with <= 2 dcs),
- there should be sufficiently many voters so that single node failures don't pose a real threat of losing majority,
- voters should be split evenly (as much as possible) among racks in a single dc.

If the requirements are less strict, then I need an explanation and I'm sorry for this long comment.

If I'm right about the requirements, then I think that achieving them consistently in all cases is a bit challenging. For example, let's ignore racks for simplicity and assume we do the following: we ensure that we have 3 voters per dc (so that we have sufficiently many voters) and if a dc has < 3 nodes, all nodes in this dc are voters. What to do if we have dc1: 2 nodes, dc2: 2 nodes, dc3: 1 node, all nodes are voters, and we suddenly add a node to dc1? We cannot make it a voter even though we want to have 3 voters per dc. We must wait until we add a node to dc2 or dc3, and then make both these nodes voters at once.

So, I expect the solution and testing to be at least a little bit complicated and without more context it's hard to judge this single test. Maybe we need something like a mini design doc with the requirements and all rules responsible for voter changes. Or maybe I just need an explanation why this task is much easier than I think.

Reply:
A different corner case: dc1: 2 nodes, dc2: 2 nodes, d3: 1 node, all nodes are voters, and we remove a node belonging to dc1. Since we remove a node, dc1 loses a voter and suddenly dc2 has half of the voters. So, apart from removing the node, we must also make one of the nodes in dc2 a non-voter.

Reply:
I was thinking about a different set of requirements.
First, there is no need to have more than 5, or 7, or 9 voters, ever. It could be a hard-coded or a configurable constant. We should also allow users to configure nodes which can never become raft voters.  Each voter in its own DC. Or, if we deal with a single or a couple DCs, in a few disticnt racks at distinct DCs.

Then we should select voters which are most distant from each other. E.g. two voters are most distant from each other if they are in different DCs and different racks. So when promoting a voter, we should find a node which is most distant from all existing voters. If a node joins the cluster that is more distant than an existing voter from other nodes, it should become a voter.

Reply:
Note that we don't need to functional-test the distance function. We can unit-test it. it's a static calculation based on node rack/dc properties.

Reply:
@patjed41 As for the number of nodes in each DC vs. the max number of voters per DC the idea was indeed that the max number of voters will se chosen so that no DC would have a majority (or a tie). Which would mean that in your example: dc1: 2 nodes, dc2: 2 nodes, dc3: 1 node and another node added to the dc1 - in this case the new node wouldn't become a voter (even though the number of nodes is lower than the voters limit).

That is in fact why I added the parametrization of [1, 3] that should also test for this use case - with the parameter of `num_nodes = 1` the DC1 would have 2 nodes and DC2 and DC3 1 node each, so the DC1 should only have 1 voter (not 2, even though the maximum allowed is higher) otherwise the test would fail (when the DC1 is shut down).

The case with nodes removal is actually an interesting one as well and we should probably also have a test for it.

Reply:
@kostja The total number of voters limit (and the distance maximization) is a very interesting idea, sounds it could work quite well. There could still be some degenerate cases (e.g. the case of DC1: 2, DC2: 1, DC:1 where the DC1 should still have just a single voter despite the limit being e.g. 5, but that could be accounted for and handled in the algo).

One downside of the "maximum distance" calculation is that it is basically a global optimization problem that might be computationally challenging ... but it might be fine with such a low number of total voters limit.

And indeed it would make sense to unit-test most of it and having just some integration test to make sure everything works together, like e.g. this one.

Reply:
So for the "degenerate" cases, where the number of nodes is very low (like in the case DC1:2, DC2:1, DC3:1) I think the requirement (condition) is that the number of voters in each DC must be `< total_nodes_count / 2`.

(for the case when we'd have a total limit of voters, e.g. 5, the condition would change to something like `< min(total_nodes_count, total_voters_count) / 2` - which would mean the DC has more priority in the distance function than the rack number; but we might also need to consider the case of different DC nodes in the same rack)

Reply:
> One downside of the "maximum distance" calculation is that it is basically a global optimization problem that might be computationally challenging ... but it might be fine with such a low number of total voters limit.

@kostja @emdroid sounds like you're overcomplicating the problem. Don't introduce a distance function, work on dc/racks directly since this is the only thing that affects "distance". When selecting a new voter, greedily prefer to select one from different DC, if not, then from different rack. That's all.

Reply:
@kbr-scylla I think we're talking about the same thing. I don't suggest having a distance matrix for every node in the cluster. But we can topologically sort all nodes, or store them in a topology tree, and find the next voter by traversing the tree from top to bottom.

I  am thinking about the following data structure:

- a tree of nodes
- the top level node (root) - cluster
- children of the cluster - dcs
- children of dcs - racks
- children of racks - nodes
- each node contains the current number of voters in all its leaves

To find a voter, we traverse the tree from root to the children, and select the path at each crossing with the least number of voters.

Reply:
@gleb-cloudius Thanks, will have a look in there to see if we can reuse some of that